library/tinkerpop/src/main/java/uk/gov/gchq/gaffer/tinkerpop/service/GafferPopNamedOperationService.java

```diff
@@ -44,44 +57,61 @@ public Type getType() {
         return Type.Start;
     }
 
+    /**
+     * Executes the Service, either calling an existing Named Operation or
+     * adding a new Named Operation.
+     *
+     * @param ctx Unused
+     * @param params {@link Map} containing a key of either "execute" or
+     * "add" and value with name to execute or operation chain to add
+     * @return {@link CloseableIterator} with results or empty if adding
+     * a Named Operation
+     */
     @Override
     public CloseableIterator<R> execute(final ServiceCallContext ctx, final Map params) {
         if (params.containsKey("execute")) {
-            final String name = (String) params.get("execute");
-            Iterable<NamedOperationDetail> namedOps = graph.execute(new OperationChain.Builder().first(new GetAllNamedOperations()).build());
-            for (final NamedOperationDetail namedOp : namedOps) {
-                if (namedOp.getOperationName().equals(name)) {
-                    if (namedOp.getOperationChainWithDefaultParams().getOutputClass().equals(Iterable.class)) {
-                        return (CloseableIterator<R>) CloseableIterator.of(graph.execute(new OperationChain.Builder()
-                                .first(new NamedOperation.Builder()
-                                        .name(name)
-                                        .build())
-                                .then(new GenerateObjects.Builder<GafferPopElement>()
-                                        .generator(new GafferPopElementGenerator(graph))
-                                        .build())
-                                .build()).iterator());
-                    } else {
-                        return CloseableIterator.of(Arrays.asList(graph.execute(new OperationChain.Builder()
-                                .first(new NamedOperation.Builder<Void, R>()
-                                        .name(name)
-                                        .build())
-                                .build())).iterator());
-                    }
-                }
-            }
-            throw new IllegalStateException("Named Operation not found");
+            return executeNamedOperation((String) params.get("execute"));
         } else if (params.containsKey("add")) {
-            final Map<String, Object> addParams = (Map) params.get("add");
-            graph.execute(new OperationChain.Builder()
-                    .first(new AddNamedOperation.Builder()
-                            .name((String) addParams.get("name"))
-                            .operationChain((String) addParams.get("opChain"))
-                            .build())
-                    .build());
-            return CloseableIterator.empty();
+            return addNamedOperation((Map<String, String>) params.get("add"));
```

Comment:
This is technically an unchecked cast is it work wrapping in a try?

Reply:
I don't think so. It's always the case that these will be strings. Potentially not if the service isn't being called correctly, but I'm not sure about that.